readme.txt
################################################################
##                                                            ##
##           The cTraceo Acoustic Raytracing Model            ##
##                                                            ##
################################################################
## Website:                                                   ##
##          https://github.com/EyNuel/cTraceo/wiki            ##
##                                                            ##
## License: This file is part of the cTraceo Raytracing Model ##
##          and is released under the Creative Commons        ##
##          Attribution-NonCommercial-ShareAlike 3.0 Unported ##
##          License (CC BY-NC-SA 3.0)                         ## 
##          http://creativecommons.org/licenses/by-nc-sa/3.0/ ##
##                                                            ##
## NOTE:    cTraceo is research code under active development.##
##          The code may contain bugs and updates are possible##
##          in the future.                                    ##
##                                                            ##  
## Written for project SENSOCEAN by:                          ##  
##          Emanuel Ey                                        ##  
##          emanuel.ey@gmail.com                              ##  
##          Copyright (C) 2011, 2012                          ##  
##          Signal Processing Laboratory                      ##  
##          Universidade do Algarve                           ##  
##                                                            ##  
## cTraceo is the C port of the FORTRAN 77 TRACEO code,       ##
## written by:                                                ##
##          Orlando Camargo Rodriguez:                        ##  
##          Copyright (C) 2010                                ##  
##          Orlando Camargo Rodriguez                         ##  
##          orodrig@ualg.pt                                   ##  
##          Universidade do Algarve                           ##  
##          Physics Department                                ##  
##          Signal Processing Laboratory                      ##  
##                                                            ##
################################################################


################################################################
## Contents:
 * Features
 * Software Requirements
 * Makefile Configuration
 * Compilation & Installation
 * Matlab Configuration & Examples
 * License Notes.

################################################################
## Features:

 # Range dependent altimetry/bathymetry including range dependent
  elastic properties (Compressional and shear wave speeds,
  density, as well as compressional and shear attenuation);
  
 # Range dependent sound speed fields;
 
 # Support for user-defined smooth objects within waveguide
   (like rocks, marine mammals, submarines);
   
 # Full suppport for backscattering (from altimetry/bathymetry
   or objects);
   
 # Allows free positioning of acoustic source and receivers
   (altimetry/bathymetry as well as objects and receivers may be
   positioned behind the source for modelling complex waveguides);
   
 # Support for horizontal, vertical, rectangular and arbitrarily
   shaped receiver arrays (allows modelling unusual array shapes,
   multiple arrays as well as array distortions);
   
 # Output options: ray trajectories and amplitudes, eigenray
   search (with user-definable proximity threshhold), arrival
   patterns, acoustic pressure, coherent transmission loss, as
   well as particle velocity components (for modelling vector
   sensors).
   
 # Outputs are written to .mat files (does not require Matlab(R)
   for compilation/usage)

################################################################
## Software Requirements

# Matlab (Optional)
  This software generates Matlab .mat-files as output, and
  provides 2 methods for writing these files.
  1 - Linking with Matlab(R) libraries during compilation.
  2 - Using a set of internal functions.
  Option 2 is now the recommended method and is enabled by default.
  Although it is still possible to fall back to option 1 this
  requires some extra configuration, is only supported on Linux
  and requires a Matlab(R) License to be available.
  

# Compiler Support
  cTraceo requires a C Compiler compatible with the ISO C99
  Standard. This is mostly due to the need for complex
  (Imaginary) Math.
  
 * The GNU C Compiler
   GCC fully supports the ISO C99 standard as of version 4.5.
   The 4.3.x and 4.4.x versions partially implement the C99
   standard and although warnings related to imaginary math may
   be generated, the code is compiled successfully.
   Successfully tested with versions 4.4.4 and 4.4.5 on Linux as
   well as with version 4.5.3 on Windows (using Cygwin)
  
 * Clang - LLVM C Language Frontend
   The Clang compiler is the compiler used for development of the
   cTraceo Model.
   Successfully tested with versions 2.8 and 2.9 on Linux.
  
 * MS Visual Studio
   No version of Microsoft's Visual Studio C Compiler supports 
   the ISO C99 Standard, so compilation should not be possible.
   Not tested.



################################################################
## Makefile Configuration
   To compile this model the user may have to adapt the included
   makefile to match the target system.
   Although the options in the Makefile are mostly self-
   -explanatory, a brief overview will be presented here:

# The "CC" variable
  Defines which command to invoke for compilation.

# The "ARCH" variable
  This is used to define whether to compile for 32 bit or 64 bit
  systems.

# The "USE_MATLAB" variable
  When compiling cTraceo on Linux it is possible to choose between
  using Matlab or the internal functions to write the result files.
  If in doubt, leave at "0". Note that linking with matlab is only
  supported in linux.
   
# The "MATLAB_DIR" variable
  This variable should contain the base directory of the 
  system's Matlab(R) installation. This is needed when linking with 
  Matlab for writing .mat files, and is only possible on Linux.
  

# The "MATLAB_VERSION" variable
  As different versions of Matlab require different variable
  types for calling library functions, this option is required 
  for selecting the appropriate types.

################################################################
## Compilation and Installation

 # Linux
   After the options descibed in "Makefile Configuration" have
   been defined, just run:
    
   $> make
   
   This will compile the model and place a binary named "ctraceo"
   in the "bin/" subdirectory.
   
 # Windows
   To compile cTraceo on windows, several steps are required:
   
    * Download and install Cygwin from www.cygwin.com.
      Cygwin provides basic linux build tools (like 'make') for windows.
      When installing Cygwin packages, select the following packages:
        "devel > gcc4"
        "devel > make"
      Several other packages will be installed for dependencies.
    
    * Add "c:/cygwin/bin" to your system's PATH environment variable.
    
    * Open Cygwin. A command line will appear.
      This command line provides an interface similar to Linux, and
      allows for easy compilation of the model.
      In Cygwin, navigate to location which contains cTraceo's source
      code; this will look similar to:
        $> cd /cygdrive/c/Users/username/Documents/models/cTraceo
    
    * Run:
        $> make
      If all goes well, this will produce an executable called
      "ctraceo.exe" in the "bin/" subdirectory which can now be
      used like any other windows executable.
    
   
################################################################
## Matlab Configuration & Example Cases
   Several example cases are contained in the "examples/"
   subdirectory. These examples depend on matlab functions
   contained in the "M-Files/" subdirectory and as such this
   folder should be added to the matlab path.
   
   Within the provided example files the model is invoked by
   calling the command "ctraceo", and thus the "bin/" folder 
   containing the compiled cTraceo binary should be on the
   system's PATH.


################################################################
## License Notes
   The cTraceo Raytracing Model is released under the Creative
   Commons Attribution-NonCommercial-ShareAlike 3.0 Unported
   (CC BY-NC-SA 3.0) 
   http://creativecommons.org/licenses/by-nc-sa/3.0/

Below, a short, "human readable" version of the software license
is shown. Read the attached "license.txt" file for the full text
of the license.
----------------------------------------------------------------

In Short:
You are free:
    to Share — to copy, distribute and transmit the work
    to Remix — to adapt the work 

Under the following conditions:
    Attribution   — You must attribute The cTraceo Raytracing
                    Model to SiPLAB (with link to
                    http://www.siplab.fct.ualg.pt/)
    Noncommercial — You may not use this work for commercial 
                    purposes. 
    Share Alike   — If you alter, transform, or build upon this
                    work, you may distribute the resulting work
                    only under the same or similar license to
                    this one. 

With the understanding that:
    Waiver        — Any of the above conditions can be waived if
                    you get permission from the copyright holder.
    Public Domain — Where the work or any of its elements is in
                    the public domain under applicable law, that
                    status is in no way affected by the license.
    Other Rights  — In no way are any of the following rights
                    affected by the license:
                    # Your fair dealing or fair use rights, or
                      other applicable copyright exceptions and
                      limitations;
                    # The author's moral rights;
                    # Rights other persons may have either in
                      the work itself or in how the work is used,
                      such as publicity or privacy rights.
    Notice        — For any reuse or distribution, you must make
                    clear to others the license terms of this
                    work. The best way to do this is with a link to:
                    http://creativecommons.org/licenses/by-nc-sa/3.0/


changelog.txt
################################################################
## Version 1.4 ##
## Features/Changes:
 # When creating a distribution package, all binaries are now statically
   linked to ensure that users can run them without compilation.
   Previous version have also been recompiled in this manner.
  
  
## Bugfixes:
 # Fixed bug #28 which caused refrection point coordinates to be stored
   incorrectly when using ERF output option.
   Thanks to Ronan Le Bras for reporting this issue.
   
 # Fixed bug #29 which caused ray amplitude values to be NaN when
   computing arrivals by Regula Falsi (ADR).
   Thanks to Ronan Le Bras for reporting this issue.
 
 # Fixed makefile's option for linking with matlab.
  
 # Undocumented input option '-' was added to the help.
 
 # Several minor bugfixes along with code and documentation
   improvements.
 
################################################################
## Version 1.3 ##
## Features/Changes:
 # Implemented an option to truncate backscattered rays. This option is
   activated by passing the command line switch '--killBackscatteredRays'.
   The number of backscattered rays is saved in the resulting .mat
   file as 'nBackscatteredRays'. This option can result in significant
   performance improvements when large amounts of backscattered rays
   are present in the waveguide.
   
 # Output file names can now be specified manually. This is done by 
   passing the command line option '--outputFileName <name>'.
  
 # Added command line switch '--noHeader' which causes cTraceo to
   write less output when being run. Usefull to reduce the amount
   of reduntant clutter when running many cases at the same time.
   
 # Added missing command line switch '--version' which causes cTraceo
   to output it's version number.
   
 # cTraceo's command line switches are no longer case-sensitive
   so that passing '--noLog' or '--nolog' is now the same.
   
 # Improved logging output.
 
 
## Bugfixes:
 # Fixed bug #19 which caused ray travel time "tau" to be slightly
   incorrectly interpolated at a hydrophone's position. 

 # Solved a bug which could cause a log file to be incorrectly named.
 
 # Solved a bug which prevented the case title from being correclty
   written to the output file.
   
 # Fixed bug #13 which caused a segfault to occurr when the source was
   placed on the range box boundary.
   
 # Several minor and "behind the scenes" code improvements.
   
################################################################
## Version 1.2 ##
## Features/Changes:
 # Added a command line option to store the interpolated sound
   speed profile, as used internally by the model, to a matfile.
   Passing '-s #' or '--ssp #' will generate an #-point interpolated
   SSP and store it in the file 'ssp.mat'.
   
 # Added a command line option to not write a log file to disk ('--noLog')
 
 # Made file access errors more verbose. The user can now clearly
   see which file caused the error.
   
 
## Bugfixes:
 # Fixed a bug which caused incorrect ssp interpolation close to
   ssp 'edges'. Linear interpolation was being used in first and
   second as well as last and second-to-last interval of ssp
   values when it should only be used in the first and last one.
   
 # Fixed a bug where the particle velocity output from a vertical
   array would be transposed.

################################################################
## Version 1.1 ##
## Features/Changes:
 # Added support for reading input files from stdin (as opposed
   to reading them from disk). This makes it possible to call
   the model from within Matlab (R) without having to write any
   files to the disk -this is done using the new "cTraceo.m"
   function contained in the M-Files/ subfolder.

 # Added support for irregularly spaced (or arbitrary) ray
   ray launching angles.
   Note that this feature required changes to the input file
   format; version 1.1 will read both input file versions
   transparently, but pre 1.1 versions do not handle the new
   input file format correclty.
 
 # Reduced the number of interpolation calls when solving dynamic 
   equations by ~50%, thus reducing overall raytracing
   time by ~5%.
 
 
## Bugfixes:
 # Fixed a bug where, when using the Regula Falsi method, the
   number arrivals [ADR] could under some conditions be different
   than the number of eigenrays [ERF].
   
 # Solved all compiler warnings when compiling with GCC.
   [tested with GCC 4.6.1]

################################################################
## Version 1.0 ##
## Features/Changes:
 # Replaced the Matlab file output API with a set of self
   written functions (see: https://github.com/EyNuel/matlabOut).
   This eliminates Matlab as a dependency, and makes compilation
   easier;
   
 # Added support for compilation on Microsoft Windows.
 
 # Removed 1D parabolic interpolation as an option for altimetry
   and bathymetry as it leads to innacurate results.
   
 # Renamed variables "rarray" and "zarray" to "arrayR" and
   "arrayZ" respectively

 # When calculating eigenrays or arrivals, more information is
   now provided. Namely: 
    iReturns:   indicates if ray inverted its horizontal direction.
    nSurRefl:   number of surface reflections
    nBotRefl:   number of bottom reflections
    nObjRefl:   number of object reflections
    nRefrac:    number of refraction points
    refrac_z:   z coordinates of refraction points
    refrac_r:   r coordinates of refraction points
    
 # Added "license.txt" to the distribution.

 # Added "manual.pdf" to the distribution.
 
 # Added help output to makefile and ctraceo executable.
 
 # Added pre-compiled binaries for Windows/Linux i686/ia64.
 
 # Added a decent header to the logfile and to the message
   printed to stdout when running the model.
 
 
## Bugfixes:
 # Fixed a bug where launched rays would be terminated to early.
 
 # Now correctly writing an "arrival" struct instead of an
   "eigenray" struct when calculating arrivals by Regula Falsi.
   

globals.h
/*
 *	globals.h
 *	Defines global variables needed by cTraceo.
*/
#pragma once
#include <stdint.h>		//needed for inequivocal type qualifiers as defined in C99
#include <stdbool.h>	//for boolean data type
#include <complex.h>


/********************************************************************************
 * Configuration:																*
 ********************************************************************************/
#ifndef VERBOSE
	#define VERBOSE					0		//when set to 1, more information will be shown.
#endif				
#define VERBOSITY					3		//verbosity level (0-10). High levels will make the code impractically slow (seriously!)
#define MAX_LINE_LEN				256		//Maximum number of chars to read at once from a file (\n not included)
#define MEM_FACTOR					20		//The memory allocated for each ray is determined like so: ( abs(rbox2-rbox1)/dsi )* MEM_FACTOR
											//NOTE:	for deepwater cases, values between 3-5 are ok.
											//		for shallow water, or other cases with a lot of reflections,
											//		values of 15-25 may be adequate
#define KEEP_RAYS_IN_MEM			0		//[boolean] determines whether a ray's coordinates are kept in memory after being written to the .mat file. (mat become usefull for multiprocessing)
#define MIN_REFLECTION_COEFFICIENT	1.0e-15	//used in solveEikonalEq(). When a rays reflection coeff is below this threshold, it is killed.



/********************************************************************************
 * ------------------------------ MATLAB --------------------------------------	*
 * Some data types vary between matlab versions, so we have to define			*
 * their type according to the matlab version used during compilation.			*
 * NOTE: To choose a specific matlab version for compilatio, edit the Makefile.	*
 ********************************************************************************/
#if USE_MATLAB == 1
	#define R12		0
	#define R14		1
	#define R2007A	2
	#define R2007B	3
	#define R2008A	4
	#define R2008B	5
	#define R2010B	6


	//When writing matlab-matrixes to matfiles, the data type varies between matlab versions.
	#if MATLAB_VERSION == R12 || MATLAB_VERSION == R14
		//matlab R14 uses int32
		#define MWSIZE			int32_t
		#define	MWINDEX			int32_t
	#elif	MATLAB_VERSION == R2007A || MATLAB_VERSION == R2007B || MATLAB_VERSION == R2008A || MATLAB_VERSION == R2008B || MATLAB_VERSION == R2010B
		//matlab R2008b and others use mwSize and mwIndex which defined in "matrix.h"
		#define MWSIZE			mwSize		
		#define MWINDEX			mwIndex
	#endif
#else
	#define MWSIZE			uintptr_t
	#define MWINDEX			uintptr_t
#endif

/********************************************************************************
 * Some utilities																*
 ********************************************************************************/

//the following 5 lines are used to simplify access to star pressure elements (for particle velocity)
#define LEFT	0
#define CENTER	1
#define RIGHT	2

#define TOP		0
#define	BOTTOM	2

#define LINUX   1
#define WINDOWS 2

//function macro used for showing debugging information:
#if VERBOSE == 1
	#define WHERESTR				"[%s,\tline %d]:\t"
	#define WHEREARG				__FILE__, __LINE__
	#define DEBUG(level, ...)		if(VERBOSITY >= level){fprintf(stderr, WHERESTR, WHEREARG);fprintf(stderr, __VA_ARGS__);}
#else
	//this should be optimized away nicely by the compiler:
	#define DEBUG(level, ...)		{}
#endif


/********************************************************************************
 * Minor data structures.														*
 ********************************************************************************/

//TODO this seems somewhat redundant...
typedef struct	vector{
	double	r;	//range component of vector
	double	z;	//depth component of vector
}vector_t;		//TODO add position components?

typedef struct	point{
	double	r;	//range component of point
	double	z;	//depth component of point
}point_t;


/********************************************************************************
 * Output data structures.														*
 *******************************************************************************/

typedef struct	ray{
	/*
	 * NOTE: memory ocupied is 44B overhead + 96B per ray coordinate. TODO recalculate, as this has since become larger
	 */
	uintptr_t		nCoords;
	bool			iKill;		//indicates if ray has been "killed"
	double			theta;		//launching angle of the ray
	double			rMin, rMax;	//used to determine if a ray "turns back"
	bool			iReturn;	//indicates if a ray "turns back"
	double*			r;			//range of ray at index
	double*			z;			//depth of ray at index
	double*			c;			//speed of sound at index
	bool*			iRefl;		//indicates if there is a reflection at a certain index of the ray coordinates.
	uint32_t		sRefl;		//number of surface reflections
	uint32_t		bRefl;		//number of bottom reflections
	uint32_t		oRefl;		//number of object reflections
	uint32_t		nRefl;		//number of total reflections
	complex double*	decay;		//decay of ray
	double*			phase;		//ray phase
	double*			tau;		//acumulated travel time
	double*			s;			//acumulated distance travelled by the ray
	double*			ic;			//see Chapter 3 of Traceo Manual
	vector_t*		boundaryTg;	//"tbdry" a boundary's tangent vector
	int32_t*		boundaryJ;	//"jbdry",	indicates at what boundary a ray is (-1 => above surface; 1 => below bottom)
	uint32_t		nRefrac;	//"nrefr", number of refraction points
	//point_t*		refrac;		//"rrefr, zrefr", coordinates of refraction points. used in "solveEikonalEq.c"
	double*			rRefrac;
	double*			zRefrac;

	double*			p;			//used in solveDynamicEq
	double*			q;			//used in solveDynamicEq
	double*			caustc;		//used in solveDynamicEq
	complex double*	amp;		//ray amplitude
}ray_t;



/********************************************************************************
 * Structures that contain the settings that are read from the input file.		*
 *******************************************************************************/
 
typedef struct source{
	double		ds;				//ray step
	double		rx,zx;			//source coords
	double		rbox1, rbox2;	//the box that limits the range of the rays
	double		freqx;			//source frequency
	uint32_t	nThetas;		//number of launching angles
	double		theta1, thetaN;	//first and last launching angle
	double		dTheta;			//the increment between launching angles
	double*		thetas;			//the array that will actually contain the launching angles (is allocated in "readin.c")
}source_t;

typedef struct interface{
	/*
		Used for both the "batimetry" as well as "altimetry" block
	*/
	//See #defines following this block for possible values
	uint32_t				surfaceType;			//formerly "atype"
	uint32_t				surfacePropertyType;	//formerly "aptype"
	double*					r;						//"rati(n)"				|
	double*					z;						//"zati(n)"				 }	these pointers are mallocced in "readin.c"
	double*					cp;						//"cpati",	compressional speed
	double*					cs;						//"csati",	shear speed
	double*					rho;					//"rhoati",	density
	double*					ap;						//"apati",	compressional attenuation
	double*					as;						//"asati"	shear attenuation
	uint32_t				surfaceInterpolation;	//formerly "aitype"
	uint32_t				surfaceAttenUnits;		//formerly "atiu"
	uint32_t				numSurfaceCoords;		//formerly "nati"
}interface_t;

//possible values for surfaceType (see page 38, Traceo manual):
#define SURFACE_TYPE__ABSORVENT	1	//formerly "A"
#define SURFACE_TYPE__ELASTIC	2	//formerly "E"
#define	SURFACE_TYPE__RIGID		3	//formerly "R"
#define SURFACE_TYPE__VACUUM	4	//formerly "V". Vacuum over surface

//possible values for surfacePropertyType (see page 38, Traceo manual):
#define	SURFACE_PROPERTY_TYPE__HOMOGENEOUS		5	//formerly "H"
#define SURFACE_PROPERTY_TYPE__NON_HOMOGENEOUS	6	//formerly "N"

//possible values for surfaceInterpolation (see page 38, Traceo manual):
#define SURFACE_INTERPOLATION__FLAT		7	//"FL", flat surface
#define SURFACE_INTERPOLATION__SLOPED	8	//"SL", sloped surface
#define SURFACE_INTERPOLATION__2P		9	//"2P", piecewise linear interpolation
#define SURFACE_INTERPOLATION__3P		10	//"3P", piecewise parabolic interpolation
#define SURFACE_INTERPOLATION__4P		11	//"4P", piecewise cubic interpolation

//possible values for attenUnits (see page 39, Traceo manual):
#define	SURFACE_ATTEN_UNITS__dBperkHz		12	//"F",	dB/kHz
#define	SURFACE_ATTEN_UNITS__dBperMeter		13	//"M",	dB/meter
#define	SURFACE_ATTEN_UNITS__dBperNeper		14	//"N",	dB/neper
#define	SURFACE_ATTEN_UNITS__qFactor		15	//"Q",	Q factor
#define	SURFACE_ATTEN_UNITS__dBperLambda	16	//"W",	dB/<lambda>


typedef struct soundSpeed{
	uint32_t	cDist;			//"cdist", type of sound speed distribution -profile or field (i.e. range dependent)
	uint32_t	cClass;			//"cclass", class of sound speed
	uint32_t	nr, nz;			//"nr0,nz0", number of points in range and depth
	double*		z;				//"z0",	depth
	double*		r;				//"r0", range
	double*		c1D;			//"c0", sound speed at (z0)
	double**	c2D;			//"c02d", sound speed at (r0,z0)
}soundSpeed_t;

//possible values for cDistribuition (see page 39, Traceo Manual)
#define	C_DIST__PROFILE		17	//"c(z,z)", sound speed profile c=c(z)
#define	C_DIST__FIELD		18	//"c(r,z)", sound speed field c=c(r,z)

//possible values for cClass	(pages 39,40, Traceo Manual)
#define	C_CLASS__ISOVELOCITY		19	//"ISOV"
#define C_CLASS__LINEAR				20	//"LINP"
#define C_CLASS__PARABOLIC			21	//"PARP"
#define C_CLASS__EXPONENTIAL		22	//"EXPP"
#define C_CLASS__N2_LINEAR			23	//"N2LP", n^2 linear
#define C_CLASS__INV_SQUARE			24	//"ISQP", inverse-square gradient
#define C_CLASS__MUNK				25	//"MUNK"
#define C_CLASS__TABULATED			26	//"TBAL"


typedef struct object{
	/*
	 * contains all information regarding a specific object.
	 */
	uint32_t				surfaceType;			//"otype",		Object surface type (rigid, aboservent, etc), as defined for interface_t
	uint32_t				surfaceAttenUnits;		//"obju",		attenuation units, as define for interface_t
	double					cp;						//"cpati",		compressional speed
	double					cs;						//"csati",		shear speed
	double					rho;					//"rhoati",		density
	double					ap;						//"apati",		compressional attenuation
	double					as;						//"asati",		shear attenuation
	uint32_t				nCoords;				//"no",			number of coordinates
	double*					r;						//"ro"		|
	double*					zDown;					//"zdn"		 >	actual coordinates that define the object geometry
	double*					zUp;					//"zup"		|
}object_t;

typedef struct objects{
	/*
	 * contains common object information as well as a pointer to an array with the actual object information.
	 */
	uint32_t	numObjects;				//"nobj"
	uint32_t	surfaceInterpolation;	//"oitype", Object interpolation type
	object_t*	object;
}objects_t;

typedef struct output{
	uint32_t			calcType;			//"catype"
	uint32_t			arrayType;			//"artype"
	uint32_t			nArrayR, nArrayZ;	//"nra", "nza"	Array sizes in R and Z
	double*				arrayR;				//"nra"			Array R (ranges)
	double*				arrayZ;				//"nrz"			Array Z (depths)
	//complex double*		pressure1D;			//will contain coherent acoustic pressure at each array element (1D), calculated in "calcCohAcoustPress.c"
	complex double**	pressure2D;			//will contain coherent acoustic pressure at each array element (2D), calculated in "calcCohAcoustPress.c"
	complex double		(**pressure_H)[3];		//used when calculating particle velocity (pressure at left, center, right)
	complex double		(**pressure_V)[3];		//used when calculating particle velocity (pressure at top, center, bottom)
												/* Note the redundancy:	the center pressure is present in both cases.
												 *						This is done for performance reasons.
												 */
	double				dr, dz;				//horizontal and vertical offset of the star pressure elements
	double				miss;				//"miss"		distance threshold for finding eigenrays
}output_t;

//possible values for calculationType (see page 43)
#define	CALC_TYPE__RAY_COORDS				27	//"RCO"
#define	CALC_TYPE__ALL_RAY_INFO				28	//"ARI"
#define	CALC_TYPE__EIGENRAYS_REG_FALSI		29	//"ERF", write Eigenrays (use Regula Falsi method)
#define	CALC_TYPE__EIGENRAYS_PROXIMITY		30	//"EPR", write Eigenrays (use Proximity method)
#define	CALC_TYPE__AMP_DELAY_REG_FALSI		31	//"ADR", write Amplitudes and Delays (use Regula Falsi method)
#define	CALC_TYPE__AMP_DELAY_PROXIMITY		32	//"ADP", write Amplitudes and Delays (use Proximity method)
#define	CALC_TYPE__COH_ACOUS_PRESS			33	//"CPR", write Coherent Acoustic Pressure
#define	CALC_TYPE__COH_TRANS_LOSS			34	//"CTL", write Coherent Transmission loss
#define	CALC_TYPE__PART_VEL					35	//"PVL", write Coherent Particle Velocity
#define	CALC_TYPE__COH_ACOUS_PRESS_PART_VEL	36	//"PAV", write Coherent Acoustic Pressure and Particle Velocity

//possible values for arrayType (Manual page 43)
#define ARRAY_TYPE__RECTANGULAR		37	//"RRY"
#define ARRAY_TYPE__HORIZONTAL		38	//"HRY"
#define ARRAY_TYPE__VERTICAL		39	//"VRY"
#define ARRAY_TYPE__LINEAR			40	//"LRY"


typedef struct settings {
	/*
	 * Contains all input information
	 */
	char*			cTitle;
	source_t		source;
	interface_t		altimetry;
	soundSpeed_t	soundSpeed;
	objects_t		objects;
	interface_t		batimetry;
	output_t		output;
}settings_t;


/********************************************************************************
 * Structures that contain the results which will be written to matfiles.		*
 *******************************************************************************/

//#define USE_MATLAB 0

#if USE_MATLAB == 1
	#include <matrix.h>		//needed for matlab matrix dimensions (mwSize)
#else
	#include "matlabOut/matlabOut.h"
#endif




typedef struct arrivals{
	/*
	 * Used in calcAmpDelxx to temporarily contain arrivals before they are written to 
	 * a matlab structure at the end of the function.
	 */
	double			nArrivals;
	mxArray*		mxArrivalStruct;
}arrivals_t;

typedef struct eigenrays{
	/*
	 * Similar to the above, but used in calcEigenrayXX.
	 * Only difference is in variable naming.
	 */
	double			nEigenrays;
	mxArray*		mxEigenrayStruct;
}eigenrays_t;
	


